### PR TITLE
Add interfaces for window package's health check source implementations.

### PR DESCRIPTION
**Pull request by jamesross3**
### Motivation
At the moment, consumers of these health check source implementations either have to take in the fully struct:
```
func NewFoo(source *window.UnhealthyIfAtLeastOneErrorSource) Foo {...}
```
... or the submit function:
```
func NewFoo(submitErrFunc func(err)) Foo {...}
```
This PR allows consumers to simply take a ErrorSubmitter of KeyedErrorSubmitter interface:
```
func NewFoo(healthCheckErrorSubmitter window.ErrorSubmitter) Foo {...}
```
This results in less leakage of the implementation of the window package's health check sources.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/132)

### CODE REVIEW

**Review by bmoylan — COMMENTED** [review]
Will wait for an ack from @splucs before merging

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/132#-:-Ltg8Jaa73XrIjD2dzoK:b-wjqi59)* status: all files reviewed, 1 unresolved discussion (waiting on @jamesross3)

___
*[status/health/window/base_source.go, line 47 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/132#-Ltg7Bmm5j7cu8Oup6dl:-Ltg7Bmn-QOQ6Bh4Fh0R:b79126k) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/19813100583afdeb5db469d97678a17e3d53e23d/status/health/window/base_source.go#L47)):*
> ```Go
> }
> 
> var _ status.HealthCheckSource = &baseHealthCheckSource{}
> ```

nit: this should check `var _ BaseHealthCheckSource = ...` instead of status.HealthCheckSource (same for the other files' type assertions)

note that this one isn't that interesting because the constructor now enforces the same thing

**Review by splucs — COMMENTED** [review]
@bmoylan this change LGTM

> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/132#-:-Ltjb5LQCGpS-zqSFFIw:b-yomnon)* status: all files reviewed, 1 unresolved discussion (waiting on @bmoylan)

___
*[status/health/window/base_source.go, line 47 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/132#-Ltg7Bmm5j7cu8Oup6dl:-LtjbG-m0G04ByxB3_QP:b2dfo9u) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/19813100583afdeb5db469d97678a17e3d53e23d/status/health/window/base_source.go#L47)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

nit: this should check `var _ BaseHealthCheckSource = ...` instead of status.HealthCheckSource (same for the other files' type assertions)

note that this one isn't that interesting because the constructor now enforces the same thing
</blockquote></details>

+1 to removing these everywhere since we have proper interfaces now.

**Comment by jamesross3** [comment]
done 👍

**Review by bmoylan — APPROVED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/132)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
